lib/flight_scheduler/persistence.rb

```diff
@@ -0,0 +1,100 @@
+#==============================================================================
+# Copyright (C) 2020-present Alces Flight Ltd.
+#
+# This file is part of FlightSchedulerController.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# FlightSchedulerController is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with FlightSchedulerController. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightSchedulerController, please visit:
+# https://github.com/openflighthpc/flight-scheduler-controller
+#==============================================================================
+
+class FlightScheduler::Persistence
+  def initialize(registry_name, filename)
+    @registry_name = registry_name
+    @path = File.join(FlightScheduler.app.config.spool_dir, filename)
+    @old_path = File.join(FlightScheduler.app.config.spool_dir, "#{filename}.old")
+  end
+
+  def load
+    unless File.exist?(@path) || File.exist?(@old_path)
+      Async.logger.info("No saved data for #{@registry_name}")
+      return nil 
+    end
+    failed = false
+    path = failed ? @old_path : @path
+    begin
+      Async.logger.info("Loading #{@registry_name} from #{path}")
+      data = JSON.load(File.open(path))
+      Async.logger.debug("Loaded data") { data }
+      data
+    rescue
+      if failed
+        raise
+      else
+        failed = true
+        retry
+      end
+    end
+  rescue
+    Async.logger.warn("Error loading #{@registry_name}: #{$!.message}")
+    raise
+  end
+
+  def save(data)
+    Async.logger.info("Saving #{@registry_name}")
+    Async.logger.debug("Serializable data") { data }
+
+    block = lambda do
+      # We jump through some hoops to make writing the save state atomic and
+      # consistent.
+      #
+      # 1. Create a copy of the original state, by creating a hard-link to it.
+      # 2. Create a tempfile, being careful to make sure we create one that
+      #    isn't automatically removed.
+      # 3. Write the content to the tempfile.
+      # 4. If all the content is written, move the tempfile to the correct
+      #    path.
+      FileUtils.mkdir_p(File.dirname(@path))
+      if File.exist?(@path)
+        FileUtils.cp_lr(@path, @old_path, remove_destination: true)
```

Comment:
I have hesitations about using hard links because they don't work on all file-systems (particularly NFS). I definitely see the need for a link, however we should try and use symbolic if possible.

I propose changing the algorithm to be:

1. Generate a new unix time-stamped path: `/base/file/path.1606841756`
2. Write the new content to the time-stamped path
3. Reform the symbolic link from the actual path to the time-stamp
4. Remove the old timestamp path

Timestamps are nice because it allows for a more advanced retention possible (e.g. keeping snapshots), however UUIDs or a montonic incrementor would work just fine.

Reply:
The limitation that hard links are only supported if both files are on the same partition is not limited to NFS.  It is a fundamental limitation of hard links.  NFS mounts support hard links as long as the exported filesystem supports hard links.

Creating a hardlink is an atomic operation.  It ensures that the backup will be exactly a copy of the main file and not a partial copy.  I believe that deleting and creating symlinks is not atomic and that there is no mechanism to edit a symlink.  So using symlinks would potentially result in a situation where the main file doesn't exist temporarily; or there is the possibility that it is a partial copy of the registry.

I may be wrong that symlinks are unsuitable.  However, I'm certain that I'm right that hard links are suitable, a.k.a. they can be used to ensure that the creation of the backup and the main file is atomic.

Your concern about the writes not being ordered is valid, but that is not related to the use hard/soft links.

The backup is in the same directory as the main file.  So it *really* ought to be on the same partition too.

Using multiple backups might be a good idea, but we then introduce a new problem of data retention.  I think that consideration is for another time.